### ISSUE TITLE
Suggestion : Implementing SkinnedModel to MonoGame core ?

### ISSUE THREAD

**Issue by Shqrdx** [mono/MonoGame#3825]
I think its good idea - many 3d games and engines using skinned models. XNA has it but as separate example - my suggestion is to include it to the MonoGame core. With AnimationPlayer, Keyframe, SkinningModelProcessor(possible SkinnedModel) as support to already existed skinned effect(and adding Skinned Effect to templates of pipeline is not a bad idea too).

**Comment by theZMan** [comment]
XNA has it as a separate example because in my experience you need to tweak your importer to get it right all the time. Its hard enough to get non skinned models to load in correctly.

If this was part of MonoGame then someone would need to spend a lot of time tweaking it for every export format out there.. I'd reccomend someone ports XNAnimation http://xnanimation.codeplex.com/ and grows that library into something MonoGame can use

**Comment by Shqrdx** [comment]
They all deprecated and for kwxport I need to install a 3dsmax11 ? My ask is about something basic and extendable solution for simple skin animation.

**Comment by theZMan** [comment]
Indeed - someone has to do the work and since those projects can't find people to work on them how will MonoGame?

But really the list was not a suggestion for solutions for you more of a reading list for anyone reading this issue who would like to take on the task you suggested.

For anyone trying to solve this problem now the XNA sample ports easily but good luck getting consistent models imported. 

Or if you are prepared to pay and therefor get support Digital Rune has a very solid engine that handles skinned models http://www.digitalrune.com/Products/GameEngine/Graphics.aspx

**Comment by Shqrdx** [comment]
Why I need to pay when I can simply use Unity or Unreal engines. They are free and much more powerful than MonoGame and Digital rune "engine". But they are expensive in term of performance and size(and harder to learn) - That why MonoGame needs a free and embedded solution for that problem. @tomspilman ? do you agreed with me ?

**Comment by tomspilman** [comment]
I wouldn't have a problem with a stock `SkinnedModel` type in MonoGame to complement the `SkinnedEffect`.  The issue is still someone has to contribute it and maintain it.

**Comment by KonajuGames** [comment]
It's similar to the `Model` class that has always been in XNA.  It's a simple implementation that does the basic job.  If you want anything more complex and you need to implement your own model class.  I think the standard `SkinnedModel` would work well in MonoGame under the same pretense.

**Comment by tomspilman** [comment]
> I think the standard `SkinnedModel` would work well in MonoGame under the same pretense.

Are you suggesting we bring in the old XNA `SkinnedModel` sample?  That could be a good starting point.

**Comment by KonajuGames** [comment]
> Are you suggesting we bring in the old XNA SkinnedModel sample? That could be a good starting point.

Yes, as @Shqrdx suggested in the original post.

**Comment by nkast** [comment]
I could submit the original SkinnedModel and then start pushing extensions for evaluations.

Some ideas I have in mind for the importer.

Optimize AnimationClip to bypass the ReflectiveReader. x2 faster loading.
(http://blog.nkast.gr/post/2014/02/02/CPU-Skinning-Faster-loading.aspx)
Combined with https://github.com/mono/MonoGame/pull/3518  this will bring it close to XNA loading times.

Instances parameter. (default=1)
This will work with normal models, add multiple copies of the Mesh and inject boneindexes (0,1,2,etc) 
Right now I do something similar in code, use GetData() to get the mesh, create a new VertexBuffer and use SkinnedEffect with 1 bone to draw multiple instances. Not all platforms support Buffer.GetData() and moving this to the Content pipeline will be faster. This can be removed once we have a working hardware instancing implementation in MG.
 
AnimationInclude { Both, MeshOnly, AnimationOnly }
Include only the Animation keyframes or only Mesh. In case you have animations on separate .fbx with the same model or multiple fbx with different LOD and same animation. 
We have done something like that in code, having a base fbx with mesh and animation, and a second fbx with higher LOD mesh without animation. We update the bones from the single  animationClip and draw one mesh or another based on camera distance.

Reorder/Group vertices based on weights
Draw vertices in 4 draw calls. vertices with 4 weights first, then 3weights, 2,1. Whether or not you see any gains from this depends on number of vertices VS cost of draw call. It might worth it if you batch all your animations or when we get cheaper draw calls (Net.Native, DX12).

**Comment by Shqrdx** [comment]
> Are you suggesting we bring in the old XNA SkinnedModel sample? That could be a good starting point.

Exactly.

@nkast  Note that XNA example (http://xbox.create.msdn.com/en-US/education/catalog/sample/skinned_model) uses GPU skinning. I dont think a CPU would be good on mobile devices. The DX11 allows much more bones than XNA to be used in shader. My suggestion also includes adding an effect template to pipeline for that.

**Comment by nkast** [comment]
@Shqrdx 
Both samples use the same AnimationClip class, the optimization work for GPU skinning too.

I think the number of bones depends on GPU capabilities, the Skinned effect set it to 72 as it was original. I should check to see if we can get that number higher but I don't hope for much. 72 bones * 16 floats really push the constant buffer to it's limit on feature level 9.3.

For the record, CPU skinning was very on WP7.x where the GPU was week and the Math class was NEON(=SIMD) optimized. I have done a couple of experiments on CPU skinning since then, like for example move it to C++/CX and writing directly to the GPU buffer instead of coping the data twice. 
It would like to revisit it some day and try a combination of  (1), what I describe above as Reorder/Group vertices based on number of bones and (2), SSE-friendly structure.
With CPU skinning you can move some work to other threads, animate at lower frame rates, pause or blend multiple clips. It can prove useful in some cases. And I think it's easy to move some pieces to GPGPU.

(1) https://molecularmusings.wordpress.com/2013/08/22/adventures-in-data-oriented-design-part-4-skinning-it-to-11/
(2) https://software.intel.com/en-us/articles/optimized-cpu-based-skinning-for-3d-games

**Comment by Shqrdx** [comment]
@nkast Thats great ! Cant wait for trying this! : )

**Comment by nkast** [comment]
I moved the SkinningSample in MonoGame. There are a couple of issues.
-The animation miss some keyframes, will investigate this.
In order to test it I had to create a SkineedEffect in code. Basically we need https://github.com/mono/MonoGame/issues/3057 fixed.

**Comment by Vodacek** [comment]
I also try to move SkinningSample to MonoGame and I got these result:
![img_11052015_204727](https://cloud.githubusercontent.com/assets/2104849/7588260/3761899c-f8bc-11e4-8eb1-285570131f5a.png)
any ideas what is wrong?

**Comment by nkast** [comment]
@Vodacek I didn't bother to test dude.fbx because of all the discussions about being old format and not supported by Assimp. Did you convert it somehow? are you sure nothing was lost? Hove you dried render it with Basic effect/no skinning?
I use one of our own mesh/animations to test it and despite a problem with clip animation data, the mesh and the few keyframes that I get are correct from what I can tell.

![testmgskinning](https://cloud.githubusercontent.com/assets/3018589/7588711/72bd63c0-f8c7-11e4-933b-ef02064dd7dc.jpg)

**Comment by Vodacek** [comment]
@nkast yep conversion was necesarry, I am using official FBX Coverter from Autodesk, When I try to open it in Cinema4D (modelling software) model and animation is ok.

**Comment by nkast** [comment]
@Vodacek  Are you drawing all MeshParts? Set WeightsPerVertex to 4?
More important, does the model draw ok with BasicEffect in MG?

**Comment by Vodacek** [comment]
yes all mesparts are drawn (I am iterating effect collections like Skinning sample does), all weightsPerVertex are set to 4, BasicEffect rendering gets me bind pose :)
Here is converted dude model for testing: http://vodacekworks.cz/dude.zip

**Comment by nkast** [comment]
That's great. I will take a look.

**Comment by mgarstenauer** [comment]
> I also try to move SkinningSample to MonoGame and I got these result: ...
any ideas what is wrong?

MonoGame uses [assimp](http://assimp.sourceforge.net/) for loading models. The problem is that the assimp does not load the Dude model correctly. (Assimp is not as robust as other libraries, such as the FBX SDK.)
Before importing your skinned models in MonoGame, try to load the model in the [assimp viewer](http://assimp.sourceforge.net/main_viewer.html). If the animations don't work in the viewer, they won't work in MonoGame. In this case the model needs to be fixed in, for example, 3dsmax. [Here](http://www.digitalrune.com/Documentation/html/6cc96ddc-4127-4ec7-889a-19cb71af0d3d.htm) are a few notes about FBX import.

**Comment by nkast** [comment]
https://github.com/mono/MonoGame/pull/3843

**Comment by nkast** [comment]
The animation I used had a few Keyframes and use Curve animation. 
Clearly the XNA importer interpolated those keyframes at some fixed interval and export a flat animation.
We have two choses here, 
do what XNA did at lower level when we read the animation data from Assimp. 
or move the code higher to the SkinnedModelProcessor and let user set the frame rate at which he wants the frame interpolated. Later we could implement interpolation in the `AnimationClip` and let curved animations pass through in the .xnb with no interpolation.

**Comment by nkast** [comment]
#3843 SkinnedModelProcessor
This is the SkinnedModel Sample with minimum changes.

#3893 Content Writer/Reader for ClipAnimation.
Branched from #3843
A  ClipAnimation  usually contains a large collection of  Keyframes . We can improve loading times significantly by no longer depend on  ReflectiveWriter<Keyframe> .
The resulting XNB is smaller since there is no need to prefix each  Keyframe  record with Reader Type. We save a byte on each record.

#3922 struct Keyframe 
Branched from #3893
Convert Keyframe to struct. Instead of allocating a large number of Keyframes and add each one on a List we now allocate a single array of structs.
The idea is to keep a low heap complexity (number of live objects & number of object references) for the GC.
Also this will improve loading times and reqired memory by a small factor.

**Comment by mgarstenauer** [comment]
> We have two choices here, 
do what XNA did at lower level when we read the animation data from Assimp. 
or move the code higher to the SkinnedModelProcessor and let user set the frame rate at which he wants the keyframes interpolated.

Let's not do what XNA did. IMHO that's just bad behavior in XNA. The model processors should be able to see the original animation data that was exported with the model.

**Comment by nkast** [comment]
I agree that XNA shouldn't interpolate key animations but are past this now. The best approach for MG is to emulate the XNA behavior  the best we can. 
For those who want to use the original keyframes (curved Animation) we can expand the Content nodes/keyframes with a Boolean to check for interpolated frames.
In the future the SkinnedModelProcessor could filter out those frames,  re-interpolate them at specific intervals (30fps/60fps) or leave them as they are. The AnimationPlayer could interpolate frames to work with those animations.

There is a property at Assimp 'AnimationBehaviour' that seems to be what we are looking for. I need to do some comparisons with XNA to see at which intervals it generates new frames, how/if it quantize time slices, etc.

**Comment by nkast** [comment]
Some more tests...
Importing an FBX with curved animation in MonoGame
clip.Duration : 00:00:10
keyframeCount : 84
bonesSetCount : 23
bones : 19, 12, 0, 7, 15, 2, 11, 6, 18, 8, 10, 9, 22, 1, 17, 16, 14, 21, 4, 5, 3, 13, 20, 
bone[0]
framesCount : 6
Time: 00:00:00
Time: 00:00:01
Time: 00:00:05.1500000
Time: 00:00:05.2666669
Time: 00:00:09.0666666
Time: 00:00:10

Importing an FBX with curved animation in XNA
clip.Duration : 00:00:10
keyframeCount : 13200
bonesSetCount : 22
bones : 16, 17, 15, 2, 14, 20, 21, 19, 1, 18, 5, 8, 9, 7, 4, 6, 12, 13, 11, 3, 10, 0, 
bone[0]
framesCount : 600
Time: 00:00:00
Time: 00:00:00.0166666
Time: 00:00:00.0333333
Time: 00:00:00.0500000
Time: 00:00:00.0666666
...
Time: 00:00:09.9500000
Time: 00:00:09.9666666
Time: 00:00:09.9833333

-XNA generates exactly 600 keyframes for every bone. 22(bones)*10(sec)*60 = 13200
while in MG bone[0] has 84 frames, bone[1] might have 4 keyframes, etc.
-XNA dropped the last frame at (Time: 00:00:10) even if it was defined in the FBX. In fact XNA dropped all Keyframes from the original animation and generated new ones at 0.01666s intervals.

We can do that in `ImportAnimation()` with something like:
```
// Interpolate frames  
                times.Clear();
                var framesPerSecond = 60.0; 
                var ticksPerFrame = aiAnimation.TicksPerSecond / framesPerSecond;
                var frames = aiAnimation.DurationInTicks / ticksPerFrame;
                for (var frame = 0.0; frame < frames; frame++)
                {
                    var time = frame * ticksPerFrame;
                    //if (!times.Contains(time))
                    //    times.Add(time);
                }
                times.Sort();
```

-I found that line https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/OpenAssetImporter.cs#L923 will round the seconds, for ex. result to .017s instead of 0.016666s. A fix would be
```
long ticks = (long)(time * (TimeSpan.TicksPerSecond / aiAnimation.TicksPerSecond));
                    channel.Add(new AnimationKeyframe(TimeSpan.FromTicks(ticks), transform));
```

We could have this fix so that Porting custom importers to/from XNA will work with no change.
Then we can provide another node based on  `AnimationContent` that will expose the keyframes and SRT Channels information with not much preprocessing.

**Comment by mgarstenauer** [comment]
> We can do that in ImportAnimation() with something like: ...

As I mentioned above, I am against adding this fix. This behavior in XNA is *undocumented*, *unexpected*, and, in my case, *unwanted*. Could be that it was not even intended by the XNA team, but rather a side product of the old FBX SDK. Correctly written model processors need to handle animation keyframes at random intervals. This change improves compatibility for poorly written XNA model importers. But it can break existing, correctly written skinned model processors for MonoGame. It at least will break parts of our model processor. :(

Resampling the input animation fits perfectly into the SkinnedModelProcessor. So, if we are going to add a SkinnedModelProcessor to MonoGame, why not show users how it is done correctly?

> I found that line ... will round the seconds, ...

Good catch. Yes, this needs to be fixed!

**Comment by nkast** [comment]
The decision here is whether we should break importers originally developed on XNA or importers based on the MG behavior. In case of FBX with flat animation the problem goes unnoticed, the problem is with FBX that have a few keyframes (CurvedAnimation). I assume it's those kind of FBX you worry about and the problem is mostly file size / memory size and you have a class to interpolate keyframes on the fly. I don't think this change 'break' your importer, but rather blow the size of .xnb, is that correct?

MG could follow the same approach, `SkinnedModelProcessor` shouldn't be a sample we point people to on how to workaround XNA incompatibilities. Rather we should improve the `AnimationClip` and work with SRT channels, for that we could expose the original SRT channel in the AnimationContent alongside with the XNA compatible/interpolated transforms/keyframes. Unfortunately that approach would require some changes to your importer. How can we do this work?
Are there any data/hints you would like to see added in the `AnimationContent`? How your engine is shipped, do you bundle a MonoGame version with it? Do you expect people to use new MG version in place? Do you offer free upgrades/patches?

Perhaps we could postpone this fix for a while and focus on the `AnimationClip` to work with both types of animations. In that case we have to provide a sample to people on how they could workaround on their XNA based content importers in the sort term. 
If the MG admins decide this is a fix they'd like to merge eventually, you could have some time to adapt your engine/importers so that wont cause problems to your users by the time a new MG version gets out.

**Comment by KonajuGames** [comment]
The XNA SkinnedModel sample was just that, a sample.  There were a lot of things it did inefficiently, and sampling the keyframes at 1/60sec intervals is one of them.  If we have a 5.0sec animation that has two keyframes, do we want to store two keyframes in the XNB (and subsequently loaded into memory), or 300 keyframes where 298 of them could have been easily calculated at runtime?

**Comment by nkast** [comment]
@KonajuGames 
 The SkinnedModelProccesor and what XNA was exposing through `AnimationContent` are two different things. In this case it is the XNA that interpolates the keyframes and hides the original SRT info from processors.
I'm sure we can keep XNA compatibility and extent the pipeline by providing more information, that would allow more advanced contentProcessors. It just requires some more work.

If XNA compatibility is not an issue here, then let's drop the keyframes & Matrix transforms entirely from the keyframes, expose just the individual SRT channels and let importers decide what they want to do with the information.

**Comment by mgarstenauer** [comment]
@nkast First, sorry if I don't answer your questions from above, even so you put a lot of thought into your response. But IMHO going into detail, would lead the discussion off track.

The goal of this issue is to create a SkinnedModelProcessor + SkinnedModel. I suggest the following solution:
- Add a SkinnedModelProcessor that exports keyframes (time + scale/rotation/translation).
- Add a SkinnedModel that linearly interpolates the keyframes at runtime.

**Comment by nkast** [comment]
@mgarstenauer Yes indeed, this discussion is about 'SkinnedModel' and we went of track. What you say will be implemented in MG and It can be independent of `ImportAnimation()` if we expose the SRT channels. IMO interpolating SRT, converting them to Matrix and decompose them later in the Content Processor is wrong.

I tried to understand why #3933 is 'unwanted' and find solutions that will work for everybody. if you no longer wish to argue about it it's understandable.

@KonajuGames I leave it up to you to decide and either merge or close #3933.
